### PR TITLE
Laboratory work #3, Anna Vorontsova - 22FPL2

### PR DESCRIPTION
**Pull request by vorontsann**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good start

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -76,6 +99,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        for token, ident in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:105]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -93,6 +122,24 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
+            return None
+
+        for token in tokenized_text:
+            self._put(token)
+
+        encoded_corpus = []
+        for token in tokenized_text:
+            if self.get_id(token) is None:
+                return None
+            else:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:139]: do you realy need `else`?

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -159,6 +238,24 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            return None
+
+        resulting_text = ""
+        for index, token in enumerate(decoded_corpus):
+            if index == 0:
+                resulting_text += token.upper()
+            elif token == self._end_of_word_token:
+                if index == len(decoded_corpus) - 1:
+                    resulting_text += "."
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:250]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -159,6 +238,24 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+                else:
+                    resulting_text += " "
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:252]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -208,6 +309,21 @@ def build(self) -> int:
...
+
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+        if not isinstance(n_grams, tuple) or n_grams is None:
+            return 1
+
+        for ngram in set(n_grams):
+            if not isinstance(ngram, tuple):
+                return 1
+            p_w_1_2 = n_grams.count(ngram)
+            p_w_1 = len([context for context in n_grams if context[:-1] == ngram[:-1]])
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:323]: count

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -270,6 +408,38 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+        if encoded_prompt is None:
+            return None
+        ngram_size = self._model.get_n_gram_size()
+
+        text = prompt
+
+        for i in range(seq_len):
+            tokens = self._model.generate_next_token(encoded_prompt[-ngram_size+1:])
+            if tokens is None:
+                return prompt + "."
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:424]: fstring

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:139]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -104,6 +151,14 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+
+        if element not in self._storage:
```
[the remark this answers is not in the document]
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:157]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -143,6 +210,18 @@ def _decode(self, corpus: tuple[int, ...]) -> Optional[tuple[str, ...]]:
...
+        if not isinstance(corpus, tuple) or len(corpus) == 0:
+            return None
+
+        decoded_corpus = []
+        for ident in corpus:
+            if not isinstance(ident, int):
+                return None
+            if self.get_token(ident) is None:
+                return None
+            decoded_corpus.append(self.get_token(ident))
```
[the remark this answers is not in the document]
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:222]: fixed

File: lab_3_generate_by_ngrams/start.py [hunk 9]
```diff
@@ -11,7 +13,13 @@ def main() -> None:
     """
     with open("./assets/Harry_Potter.txt", "r", encoding="utf-8") as text_file:
         text = text_file.read()
-    result = None
+        text_processor = main_py.TextProcessor('_')
```
[the remark this answers is not in the document]
Comment (vorontsann) [reply lab_3_generate_by_ngrams/start.py:16]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:250]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:252]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:424]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -270,6 +408,38 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+            for candidate, freq in tokens.items():
+                if freq == max_freq:
+                    max_candidates.append(candidate)
+            encoded_prompt = encoded_prompt + (sorted(max_candidates)[0])
+            best_candidate = self._text_processor.get_token(encoded_prompt[-1])
+            if best_candidate is None:
+                return None
+            text += best_candidate
+
+        decoded_prompt = self._text_processor.decode(encoded_prompt) + "."
```
[the remark this answers is not in the document]
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:438]: fixed??? if you meant fstring

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -41,6 +44,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        tokenized_text = []
+        for element in text.lower():
+            if element.isalpha():
+                tokenized_text.append(element)
+            elif element.isspace() and tokenized_text[-1] != self._end_of_word_token:
+                tokenized_text.append(self._end_of_word_token)
+        if not text[-1].isalnum():
+            tokenized_text.append(self._end_of_word_token)
+
+        if len(tokenized_text) == 0:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:59]: `not tokenized_text`

File: lab_3_generate_by_ngrams/main.py [hunk 12]
```diff
@@ -104,6 +151,13 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1 or element in self._storage:
+            return None
+
+        self._storage[element] = len(self._storage)
+
+        return None
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:159]: There is a way to fix mypy/pylint

File: lab_3_generate_by_ngrams/main.py [hunk 13]
```diff
@@ -159,6 +238,23 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+
+        resulting_text = ""
+        for index, token in enumerate(decoded_corpus):
+            if index == 0:
+                resulting_text += token.upper()
+            elif token == self._end_of_word_token:
+                if index == len(decoded_corpus) - 1:
+                    resulting_text = f"{resulting_text}."
+                else:
+                    resulting_text = f"{resulting_text} "
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:252]: save the `end_of_the_text` into a variable and only then, after if/else format the string with `end_of_the_text`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 11 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:59]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 13 above)
Comment (vorontsann) [reply lab_3_generate_by_ngrams/main.py:252]: fixed

**Pull request closed without merging**